_posts/2018-02-15-nohoist.md

```diff
@@ -0,0 +1,212 @@
+
+* can't find module "B@2.0" from project root "monorepo" (not able to follow symlink)
+* can't find module "A@1.0" from "package-1" (unaware of the module tree above in "monorepo")
+
+For this monorepo project to reliably find any module from anywhere, it needs to traverse each node*modules tree: *"monorepo/node*modules"* and _"monorepo/packages/package-1/node_modules"_ .
+
+## Why can't they be fixed?
+
+There are indeed many ways library owners can address these issues, such as multi-root, custom module map, clever traversing scheme, among others... However,
+
+1. not all the 3rd party libraries have the resource to adapt for monorepo environment
+1. the weakest link problem: javascript is great thanks to the massive 3rd-party libraries. However, that also means the complex tool chain is only as strong as the weakest link. A single non-adapted package deep down the tool chain could render the whole tool useless.
+1. the bootstrap problems: for example, react-native has provided a way to configure multi-root through `rn-cli.config.js`. But it won't help the bootstrap process like `react-native init` or `create-react-native-app`, which has no access of any such tooling before the app is created/installed.
+
+It is frustrating when a solution worked for a standalone project only fell short in the monorepo environment. The ideal solution lies in addressing the underlying libraries, but the reality is far from perfect and we all know that our projects can't wait...
+
+## What is "nohoist" ?
+
+Is there a simple yet universal mechanism that can allow these incompatible libraries working in the monorepo environment?
+
+Turns out there is, and is conveniently called _"nohoist"_, which has also been demonstrated in other monorepo tools like [lerna](https://github.com/lerna/lerna/blob/master/doc/hoist.md).
+
+"nohoist" enables workspaces to consume 3rd-party libraries not yet compatible with its hoisting scheme. The idea is to disable the selected modules from being hoisted to the project root. They were placed in the actual (child) project instead, just like in a standalone, non-workspaces, project.
+
+Since most 3rd-party libraries already worked in standalone projects, the ability to simulate such environment within workspaces should be able to unblock many known compatibility issues.
+
+### A word of caution
+
+While nohoist is useful, it does come with drawbacks. The most obvious one is the nohoist modules could be duplicated in multiple locations, denying the benefit of hoisting mentioned above. Therefore, we recommend to keep nohoist scope as small and explicit as possible in your project.
+
+## When will it be available?
+
+[#4979](https://github.com/yarnpkg/yarn/pull/4979) was merged on 1/29/2018. According to the [core team](https://github.com/yarnpkg/yarn/pull/4979#issuecomment-364486016), it should be deployed with **1.4.2** in mid Feb...
```

Comment:
We can release this blog post at the same time as the release I guess

Reply:
ok, updated the wording.